controllers/liveupdate_controller.go

```diff
@@ -0,0 +1,653 @@
+		return reconcile.Result{}, err
+	}
+
+	// Delete old statefulSet
+	sourceStatefulSet := &appsv1.StatefulSet{}
+	err = r.Get(context.TODO(), types.NamespacedName{Namespace: ispn.Namespace, Name: ispn.Name}, sourceStatefulSet)
+
+	if err != nil && errors.IsNotFound(err) {
+		r.reqLogger.Error(err, "source stateful set not found")
+		return reconcile.Result{}, err
+	}
+
+	err = r.Delete(r.ctx, sourceStatefulSet)
+	if err != nil {
+		r.reqLogger.Error(err, "Error Deleting statefulSet")
+		return reconcile.Result{}, err
+	}
+
+	// Remove any LiveUpdate status from the Infinispan CR
+	_, err = controllerutil.CreateOrPatch(r.ctx, r, ispn, func() error {
+		ispn.Status.LiveUpgradeStage = ""
+		return nil
+	})
+
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("error removing live upgrade status from Infinispan")
+	}
+
+	r.reqLogger.Info("Live Update finished successfully")
+	return ctrl.Result{}, nil
+}
+
+// updateStage Updates the status of the Live Update process
+func (r *LiveUpdateRequest) updateStage(stage ispnv1.LiveUpgradeStage) (reconcile.Result, error) {
+	_, err := controllerutil.CreateOrPatch(r.ctx, r, r.infinispan, func() error {
+		if r.infinispan.CreationTimestamp.IsZero() {
+			return errors.NewNotFound(schema.ParseGroupResource("infinispan.infinispan.org"), r.infinispan.Name)
+		}
+		r.infinispan.Status.LiveUpgradeStage = stage
+		return nil
+	})
+	return reconcile.Result{}, err
+}
+
+// generateTargetClusterName Derive a name for the new statefulSet.
+func generateTargetClusterName(sourceName string) string {
+	const (
+		Separator = "-"
+		Pattern   = Separator + "[\\d]+$"
+		Suffix    = "-1"
+	)
+	match, _ := regexp.Match(Pattern, []byte(sourceName))
+	if match {
+		split := strings.Split(sourceName, Separator)
+		version, _ := strconv.Atoi(split[1])
+		version++
+		return split[0] + Separator + strconv.Itoa(version)
+	}
+
+	return sourceName + Suffix
+}
+
+// reconcileNewPingService Reconcile a specific ping service for the pods of the new statefulSet
+func (r *LiveUpdateRequest) reconcileNewPingService() error {
+	namespace := r.infinispan.GetNamespace()
+	pingService := &corev1.Service{}
+	newPingServiceName := fmt.Sprintf("%s-ping", r.targetName)
+	err := r.Get(r.ctx, types.NamespacedName{Namespace: namespace, Name: newPingServiceName}, pingService)
+
+	if err != nil && errors.IsNotFound(err) {
+		// Obtains the existing ping service
+		pingService := &corev1.Service{}
+		err = r.Get(r.ctx, types.NamespacedName{Namespace: namespace, Name: r.infinispan.GetPingServiceName()}, pingService)
+		if err != nil {
+			r.reqLogger.Error(err, fmt.Sprintf("Could not find ping service '%s'", r.infinispan.GetPingServiceName()))
+			return err
+		}
+		// Clone existing service changing the DNS Query
+		newPingService := pingService.DeepCopy()
+		labels := newPingService.Labels
+		newPingService.ObjectMeta.ResourceVersion = ""
+		newPingService.Name = newPingServiceName
+		newPingService.Labels = labels
```

Comment:
I'm not sure why this is necessary? It seems that `labels` is already `newPingService.Labels`. Is this supposed to be `labels := pingService.Labels`?

No need for `labels` var, assignment can be direct to `newPingService.Labels = pingService.Labels`.